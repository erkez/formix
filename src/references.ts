import { List, Seq } from 'immutable';

import FieldStateImpl from './FieldState';
import { ArrayFieldRef, FieldRef, FieldState, GenericFieldRef } from './types';

class FieldRefImpl<V> implements FieldRef<V> {
    readonly initialState: FieldState<V>;

    constructor(readonly initialValue: V, initialDisabled: boolean) {
        this.initialState = new FieldStateImpl(initialValue, initialDisabled, false, false);
    }
}

export function defineField<V>(initialValue: V, initialDisabled = false): FieldRef<V> {
    return new FieldRefImpl(initialValue, initialDisabled);
}

export function isFieldRef(field: unknown): field is FieldRef<unknown> {
    return field instanceof FieldRefImpl;
}

class ArrayFieldRefImpl<T extends GenericFieldRef, V>
    extends FieldRefImpl<List<T>>
    implements ArrayFieldRef<T, V> {
    constructor(
        initialItems: List<T>,
        readonly itemTemplate: (value: V) => T,
        initialDisabled: boolean
    ) {
        super(initialItems, initialDisabled);
    }
}

export function defineArrayField<T extends GenericFieldRef, V>(
    initialItems: Iterable<V>,
    itemTemplate: (value: V) => T,
    initialDisabled = false
): ArrayFieldRef<T, V> {
    return new ArrayFieldRefImpl(
        Seq(initialItems).map(itemTemplate).toList(),
        itemTemplate,
        initialDisabled
    );
}

export function isArrayFieldRef(field: unknown): field is ArrayFieldRef<GenericFieldRef, unknown> {
    return field instanceof ArrayFieldRefImpl;
}
