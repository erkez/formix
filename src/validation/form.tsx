import { Seq, List, Map, OrderedSet } from 'immutable';
import * as React from 'react';

import { isArrayFieldRef, isFieldRef } from '../references';
import {
    GenericFieldRef,
    FieldRef,
    ValidationResult,
    FormApi,
    FormValidationApi,
    FormValidations,
    FieldValidation
} from '../types';

export function useFormValidation(getFieldState: FormApi['getFieldState']): FormValidationApi {
    const [validations, setValidations] = React.useState<FormValidations>(Map);

    const registerValidation: FormValidationApi['registerValidation'] = React.useCallback(
        (validation) => {
            setValidations((previous) =>
                previous.update(validation.fieldRef, OrderedSet(), (validations) =>
                    validations.add(validation)
                )
            );
            return () =>
                setValidations((previous) =>
                    previous.update(validation.fieldRef, OrderedSet(), (validations) =>
                        validations.remove(validation)
                    )
                );
        },
        []
    );

    const getValidationResult: FormValidationApi['getValidationResult'] = React.useCallback(
        (fieldRef) => {
            return validations
                .get(fieldRef, OrderedSet<FieldValidation<unknown>>())
                .first(EmptyValidationResult);
        },
        [validations]
    );

    const isFieldValid: FormValidationApi['isFieldValid'] = React.useCallback(
        (fieldRef) => {
            function flatten(field: GenericFieldRef): List<FieldRef<unknown>> {
                if (isArrayFieldRef(field)) {
                    return getFieldState(field).value.flatMap(flatten);
                } else if (isFieldRef(field)) {
                    return List.of(field);
                } else {
                    return Seq.Keyed(field).valueSeq().flatMap(flatten).toList();
                }
            }

            const fields = flatten(fieldRef).toSet();

            return validations
                .filter((_, field) => fields.has(field))
                .valueSeq()
                .flatMap((v) => v)
                .every((v) => v.error == null && !v.pending);
        },
        [getFieldState, validations]
    );

    const pending = React.useMemo(() => {
        return validations
            .valueSeq()
            .flatMap((v) => v)
            .some((v) => v.pending);
    }, [validations]);

    const valid = React.useMemo(() => {
        return (
            !pending &&
            validations
                .valueSeq()
                .flatMap((v) => v)
                .every((v) => v.error == null)
        );
    }, [pending, validations]);

    return { registerValidation, getValidationResult, isFieldValid, valid, pending };
}

const EmptyValidationResult: ValidationResult = Object.freeze({
    error: null,
    pending: false
});
