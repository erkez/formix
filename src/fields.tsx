import * as React from 'react';

import { useForm } from './form';
import { ArrayField, ArrayFieldRef, Field, FieldRef, GenericFieldRef } from './types';

export function useField<T>(fieldRef: FieldRef<T>): Field<T> {
    const { getFieldState, setFieldState, getValidationResult } = useForm();

    const fieldState = getFieldState(fieldRef);
    const validationResult = getValidationResult(fieldRef);

    const setValue = React.useCallback(
        (value: React.SetStateAction<T>) => {
            setFieldState<T>(fieldRef, (previousState) => {
                const nextValue: T =
                    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
                    // @ts-ignore
                    typeof value === 'function' ? value(previousState.value) : value;

                return previousState.value === nextValue
                    ? previousState
                    : previousState.updated({
                          value: nextValue,
                          touched: true
                      });
            });
        },
        [fieldRef, setFieldState]
    );

    const setDisabled = React.useCallback(
        (disabled) => setFieldState(fieldRef, (state) => state.updated({ disabled })),
        [fieldRef, setFieldState]
    );

    const onFocus = React.useCallback(
        () => setFieldState(fieldRef, (state) => state.updated({ active: true, touched: true })),
        [fieldRef, setFieldState]
    );

    const onBlur = React.useCallback(
        () => setFieldState(fieldRef, (state) => state.updated({ active: false })),
        [fieldRef, setFieldState]
    );

    return React.useMemo(
        () => ({
            value: fieldState.value,
            ...validationResult,
            disabled: fieldState.disabled,
            active: fieldState.active,
            touched: fieldState.touched,
            setValue,
            setDisabled,
            onFocus,
            onBlur
        }),
        [fieldState, validationResult, setValue, setDisabled, onFocus, onBlur]
    );
}

export function useArrayField<T extends GenericFieldRef, V>(
    fieldRef: ArrayFieldRef<T, V>
): ArrayField<T, V> {
    const { getValidationResult } = useForm();

    const {
        value: items,
        setValue,
        disabled,
        active,
        touched,
        setDisabled,
        onBlur,
        onFocus
    } = useField(fieldRef);

    const validationResult = getValidationResult(fieldRef);

    const unshift = React.useCallback(
        (value: V) => {
            const item = fieldRef.itemTemplate(value);
            setValue((previous) => previous.unshift(item));
            return item;
        },
        [setValue, fieldRef]
    );

    const push = React.useCallback(
        (value: V) => {
            const item = fieldRef.itemTemplate(value);
            setValue((previous) => previous.push(item));
            return item;
        },
        [setValue, fieldRef]
    );

    const insert = React.useCallback(
        (index, value) => {
            const item = fieldRef.itemTemplate(value);
            setValue((items) => items.insert(index, item));
            return item;
        },
        [setValue, fieldRef]
    );

    const remove = React.useCallback(
        (value: T) => {
            setValue((previous) => {
                const index = previous.indexOf(value);
                return index !== -1 ? previous.remove(index) : previous;
            });
        },
        [setValue]
    );

    const swap = React.useCallback(
        (indexA, indexB) => {
            setValue((previous) => {
                const itemA = previous.get(indexA);
                const itemB = previous.get(indexB);

                if (itemA == null || itemB == null) {
                    throw new Error('Invalid element access');
                }

                return previous
                    .remove(indexA)
                    .insert(indexA, itemB)
                    .remove(indexB)
                    .insert(indexB, itemA);
            });
        },
        [setValue]
    );

    return React.useMemo(
        () => ({
            items,
            ...validationResult,
            unshift,
            push,
            insert,
            remove,
            swap,
            disabled,
            active,
            touched,
            setDisabled,
            onBlur,
            onFocus
        }),
        [
            items,
            validationResult,
            unshift,
            push,
            insert,
            remove,
            swap,
            disabled,
            active,
            touched,
            setDisabled,
            onBlur,
            onFocus
        ]
    );
}

type UseFieldProps<T> = {
    field: FieldRef<T>;
    render(field: Field<T>): React.ReactNode;
};

export function UseField<T>(props: UseFieldProps<T>): React.ReactElement {
    const field = useField(props.field);
    return <React.Fragment>{props.render(field)}</React.Fragment>;
}

type UseArrayFieldProps<T extends GenericFieldRef, V> = {
    field: ArrayFieldRef<T, V>;
    render(field: ArrayField<T, V>): React.ReactNode;
};

export function UseArrayField<T extends GenericFieldRef, V>(
    props: UseArrayFieldProps<T, V>
): React.ReactElement {
    const field = useArrayField(props.field);
    return <React.Fragment>{props.render(field)}</React.Fragment>;
}
