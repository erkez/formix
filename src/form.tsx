import { Map, Seq } from 'immutable';
import * as React from 'react';

import { isArrayFieldRef, isFieldRef } from './references';
import { GenericFieldRef, FieldRef, FieldState, FormApi, FormWithFields } from './types';
import { useFormValidation } from './validation/form';

function useFormApi(): FormApi {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const [fieldStates, setFieldStates] = React.useState<Map<FieldRef<any>, FieldState<any>>>(Map);

    const getFieldState = React.useCallback(
        function getFieldState<T>(fieldRef: FieldRef<T>) {
            return fieldStates.get(fieldRef, fieldRef.initialState);
        },
        [fieldStates]
    );

    const getFieldValue = React.useCallback<FormApi['getFieldValue']>(
        (fieldRef) => getFieldState(fieldRef).value,
        [getFieldState]
    );

    const setFieldState = React.useCallback(function setFieldState<T>(
        fieldRef: FieldRef<T>,
        state: React.SetStateAction<FieldState<T>>
    ) {
        setFieldStates((xs) =>
            typeof state === 'function'
                ? xs.update(fieldRef, fieldRef.initialState, state)
                : xs.set(fieldRef, state)
        );
    }, []);

    const extract = React.useCallback(
        (fields: GenericFieldRef, type: 'value' | 'state') => {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            function extractField(field: GenericFieldRef): any {
                if (isArrayFieldRef(field)) {
                    return getFieldState(field).value.map(extractField);
                } else if (isFieldRef(field)) {
                    const state = getFieldState(field);
                    return type === 'value' ? state.value : state;
                } else {
                    return Seq.Keyed(field)
                        .map((value) => extractField(value))
                        .toObject();
                }
            }

            return extractField(fields);
        },
        [getFieldState]
    );

    const getStates: FormApi['getStates'] = React.useCallback(
        (fields) => extract(fields, 'state'),
        [extract]
    );

    const getValues: FormApi['getValues'] = React.useCallback(
        (fields) => extract(fields, 'value'),
        [extract]
    );

    const reset = React.useCallback(() => setFieldStates(Map()), [setFieldStates]);

    const modified = React.useMemo(() => {
        return fieldStates.some((state, fieldRef) => !state.equals(fieldRef.initialState));
    }, [fieldStates]);

    const formValidation = useFormValidation(getFieldState);

    return React.useMemo(
        () => ({
            getFieldState,
            getFieldValue,
            setFieldState,
            getStates,
            getValues,
            reset,
            modified,
            ...formValidation
        }),
        [
            getFieldState,
            getFieldValue,
            setFieldState,
            getStates,
            getValues,
            reset,
            modified,
            formValidation
        ]
    );
}

function noop(): never {
    throw new Error('Cannot call method on empty context. Did you forget <Form />?');
}

const FormContext = React.createContext<FormApi>({
    getFieldState: noop,
    getFieldValue: noop,
    setFieldState: noop,
    getStates: noop,
    getValues: noop,
    getValidationResult: noop,
    registerValidation: noop,
    isFieldValid: noop,
    reset: noop,
    pending: false,
    valid: false,
    modified: false
});

export function FormProvider(props: { children: React.ReactNode }): React.ReactElement {
    const api = useFormApi();
    return <FormContext.Provider value={api}>{props.children}</FormContext.Provider>;
}

export function useForm(): FormApi {
    return React.useContext(FormContext);
}

export function useFormWithFields<Deps extends unknown[], F extends GenericFieldRef>(
    fieldsCreator: (...deps: [...Deps]) => F,
    deps: [...Deps]
): FormWithFields<F> {
    const { reset, ...form } = useForm();

    const fields = React.useMemo(() => {
        return fieldsCreator(...deps);
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, deps);

    React.useLayoutEffect(() => {
        reset();
    }, [reset, fields]);

    return React.useMemo(
        () => ({
            ...form,
            reset,
            fields
        }),
        [form, reset, fields]
    );
}

export function withForm<Props>(Component: React.ComponentType<Props>): React.ComponentType<Props> {
    return function WithForm(props: Props): React.ReactElement {
        return (
            <FormProvider>
                {/* eslint-disable-next-line @typescript-eslint/ban-ts-comment */}
                {/* @ts-ignore-error */}
                <Component {...props} />
            </FormProvider>
        );
    };
}

type UseFormProps = {
    children: (form: FormApi) => React.ReactNode;
};

export function UseForm(props: UseFormProps): React.ReactElement {
    const form = useForm();
    return <React.Fragment>{props.children(form)}</React.Fragment>;
}
