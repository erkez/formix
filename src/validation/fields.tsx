import * as React from 'react';

import { useForm } from '../form';
import {
    GenericFieldRef,
    FormValues,
    FormStates,
    FieldRef,
    FieldRefValues,
    FieldRefStates,
    ValidationError,
    ValidationResult
} from '../types';

export function useFieldValidation<F extends FieldRef<unknown>, D extends GenericFieldRef[] = []>(
    fieldRef: F,
    validate: (value: FormValues<F>, ds: [...FieldRefValues<D>]) => ValidationError,
    deps?: [...D]
): void {
    const { getValues, registerValidation } = useForm();

    const value = getValues(fieldRef);

    const dependentValues = React.useMemo(() => {
        return (deps ?? []).map(getValues) as [...FieldRefValues<D>];
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [getValues, ...(deps ?? [])]);

    const error = React.useMemo(() => {
        return validate(value, dependentValues);
        // validate should not be considered for changes
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [value, ...dependentValues]);

    const validation = React.useMemo(
        () => ({
            fieldRef,
            error,
            pending: false
        }),
        [fieldRef, error]
    );

    React.useLayoutEffect(() => registerValidation(validation), [registerValidation, validation]);
}

function useAsyncValidationResult(
    validate: () => Promise<ValidationError>,
    debounceMs: number,
    deps: React.DependencyList
): ValidationResult {
    const [result, setResult] = React.useState<ValidationResult>({
        pending: false,
        error: null
    });

    React.useEffect(() => {
        let mounted = true;

        setResult((previous) => ({ pending: true, error: previous.error }));

        const timeoutId = setTimeout(() => {
            validate()
                .then((error) => {
                    if (mounted) {
                        setResult({ pending: false, error: error });
                    }
                })
                .catch((error) => {
                    if (mounted) {
                        setResult({ pending: false, error: 'Unexpected error' });
                    }
                    // eslint-disable-next-line no-console
                    console.warn('[Unhandled Validation Error]', error);
                    // eslint-disable-next-line no-console
                    console.warn(
                        'An async validator should never throw, but it should resolve with the validation result.'
                    );
                });
        }, debounceMs);

        return () => {
            mounted = false;
            clearTimeout(timeoutId);
        };
        // validate should not be considered for changes
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [debounceMs, ...deps]);

    return result;
}

export function useAsyncFieldValidation<
    F extends FieldRef<unknown>,
    D extends GenericFieldRef[] = []
>(
    fieldRef: F,
    validate: (value: FormValues<F>, ds: [...FieldRefValues<D>]) => Promise<ValidationError>,
    deps?: [...D],
    debounceMs = 200
): void {
    const { getValues, registerValidation } = useForm();

    const value = getValues(fieldRef);

    const dependentValues = React.useMemo(() => {
        return (deps ?? []).map(getValues) as [...FieldRefValues<D>];
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [getValues, ...(deps ?? [])]);

    const result = useAsyncValidationResult(
        () => {
            return validate(value, dependentValues);
        },
        debounceMs,
        [value, ...dependentValues]
    );

    const validation = React.useMemo(
        () => ({
            fieldRef,
            error: result.error,
            pending: result.pending
        }),
        [fieldRef, result.error, result.pending]
    );

    React.useLayoutEffect(() => registerValidation(validation), [registerValidation, validation]);
}

export function useFieldStateValidation<
    F extends FieldRef<unknown>,
    D extends GenericFieldRef[] = []
>(
    fieldRef: F,
    validate: (
        fieldState: FormStates<F>,
        dependentFieldStates: [...FieldRefStates<D>]
    ) => ValidationError,
    deps?: [...D]
): void {
    const { getStates, registerValidation } = useForm();

    const fieldState = getStates(fieldRef);

    const dependentStates = React.useMemo(() => {
        return (deps ?? []).map(getStates) as [...FieldRefStates<D>];
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [getStates, ...(deps ?? [])]);

    const error = React.useMemo(() => {
        return validate(fieldState, dependentStates);
        // validate should not be considered for changes
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [fieldState, ...dependentStates]);

    const validation = React.useMemo(
        () => ({
            fieldRef,
            error,
            pending: false
        }),
        [fieldRef, error]
    );

    React.useLayoutEffect(() => registerValidation(validation), [registerValidation, validation]);
}

export function useAsyncFieldStateValidation<
    F extends FieldRef<unknown>,
    D extends GenericFieldRef[] = []
>(
    fieldRef: F,
    validate: (fieldState: FormStates<F>, ds: [...FieldRefStates<D>]) => Promise<ValidationError>,
    deps?: [...D],
    debounceMs = 200
): void {
    const { getStates, registerValidation } = useForm();

    const fieldState = getStates(fieldRef);

    const dependentValues = React.useMemo(() => {
        return (deps ?? []).map(getStates) as [...FieldRefStates<D>];
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [getStates, ...(deps ?? [])]);

    const result = useAsyncValidationResult(
        () => {
            return validate(fieldState, dependentValues);
        },
        debounceMs,
        [fieldState, ...dependentValues]
    );

    const validation = React.useMemo(
        () => ({
            fieldRef,
            error: result.error,
            pending: result.pending
        }),
        [fieldRef, result.error, result.pending]
    );

    React.useLayoutEffect(() => registerValidation(validation), [registerValidation, validation]);
}
