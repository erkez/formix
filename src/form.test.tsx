import '@testing-library/jest-dom/extend-expect';
import { fireEvent, render } from '@testing-library/react';
import * as React from 'react';

import { useField } from './fields';
import { FormProvider, useForm, useFormWithFields, withForm } from './form';
import { defineField } from './references';

describe('form', () => {
    it('should render an empty FormProvider', () => {
        const { container } = render(
            <FormProvider>
                <form className="some-form" />
            </FormProvider>
        );
        expect(container).toMatchSnapshot();
    });

    it('should render a FormProvider and extract the form API', () => {
        const Form = withForm(function Form() {
            const form = useForm();
            return (
                <div>
                    <div>valid: {form.valid && 'YES'}</div>
                    <div>pending: {!form.pending && 'NO'}</div>
                    <div>modified: {!form.modified && 'NO'}</div>
                </div>
            );
        });

        const { container } = render(<Form />);
        expect(container).toMatchSnapshot();
    });

    it('should render a FormProvider with fields', async () => {
        const Form = withForm(function Form() {
            const { fields } = useFormWithFields(
                () => ({
                    name: defineField('fizz')
                }),
                []
            );

            const nameField = useField(fields.name);

            return (
                <div>
                    <div data-testid="name">{nameField.value}</div>
                    <input
                        value={nameField.value}
                        onChange={(e) => nameField.setValue(e.target.value)}
                    />
                </div>
            );
        });

        const { findByTestId, findByDisplayValue } = render(<Form />);
        const input = await findByDisplayValue('fizz');
        fireEvent.change(input, { target: { value: 'buzz' } });
        const nameElement = await findByTestId('name');
        expect(nameElement).toHaveTextContent('buzz');
    });
});
