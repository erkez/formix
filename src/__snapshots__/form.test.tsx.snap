// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form should render a FormProvider and extract the form API 1`] = `
<div>
  <div>
    <div>
      valid: 
      YES
    </div>
    <div>
      pending: 
      NO
    </div>
    <div>
      modified: 
      NO
    </div>
  </div>
</div>
`;

exports[`form should render an empty FormProvider 1`] = `
<div>
  <form
    class="some-form"
  />
</div>
`;
